Dr. Elephant Tez Support working patch

chinmayms opened the pull request:
Dr. Elephant Version with support for Tez based on patch (PR#278).

Summary:
Working Version with Tez Support currently running on Electronic Arts Production Data Infrastructure.
- Added 10 heuristics to analyze and flag Tez jobs.
- Added Resources used and Resources wasted logic based on MR version.
- Added 'flow history' and 'job history' graph support for Tez.
- Changed logic from choosing the first DAG to considering DAG with highest number of tasks in case of multiple DAGs in one application.
- Replicated mapreduce test suite for Tez fetcher, heuristics and data structure classes. 

Future Work:
Fundamentally one Tez application when returned by RM may contain multiple DAGs, the original Dr. Elephant Interfaces do not allow for us to go any lower than application level as is the case for MapReduce. Need design changes to further allow Vertex, Edge level analysis.

shankar37 commented on app-conf/FetcherConf.xml:
```diff
@@ -29,6 +29,13 @@
 -->
 <fetchers>
+  <!--
+     REST based fetcher for Tez jobs which pulls job metrics and data from Timeline Server API
```
Call this dependency in the docs and setup instructions.

chinmayms commented:
@akshayrai , thanks for your review. Addressing your issues one by one. Let me know your thoughts.
1) The heuristics use the same logic as the original MR ones, however, while implementing the Heuristic interface, they internally use "Tez" specific data structures and classes in the "apply" function. We could make design level changes to have standard heursitics for both but even then they would need to implement Tez specific features at some point. I am ready to take this effort forward in future versions if you think it's a good idea.
2) Applied Apache license for all files.
3) Converted Indent to 2 space.
4) Will look for Tez specific changes once @shkhrgpt completes his review.

shkhrgpt commented on app/com/linkedin/drelephant/tez/fetchers/TezFetcher.java:
```diff
@@ -0,0 +1,480 @@
+
+  public TezFetcher(FetcherConfigurationData fetcherConfData) throws IOException {
+    this._fetcherConfigurationData = fetcherConfData;
+    final String applicationHistoryAddr = new Configuration().get("yarn.timeline-service.webapp.address");
```
Declare yarn.timeline-service.webapp.address as a private static final variable.

shkhrgpt commented on app/com/linkedin/drelephant/tez/fetchers/TezFetcher.java:
```diff
@@ -0,0 +1,481 @@
+  }
+}
+
+final class ThreadContextMR2 {
```
ThreadContextMR2 is also used by MapReduceFetcherHadoop2. Refactor code here so there is only one instance of  ThreadContextMR2.

shkhrgpt commented on app/com/linkedin/drelephant/tez/fetchers/TezFetcher.java:
```diff
@@ -0,0 +1,481 @@
+      }
+    }
+
+    if(mapperListAggregate.isEmpty() && reducerListAggregate.isEmpty()){
```
Why not move this check to after line 113 with a something like 
```
if (state.equals("FAILED") { 
      jobData.setSucceeded(false); 
}
```

chinmayms commented:
@ray-harrison 
Sure, sounds good.
Can we do it on Tuesday 03/27 or Wednesday 03/28 between 1:30pm to 3:30 pm?
Let me know if these don't work for you.

chrevanthreddy commented:
@chinmayms Can we do it on Wednesday please. 1:30PM EST? I will send an invite accordingly.

shkhrgpt commented:
@ray-harrison, and @chrevanthreddy Thanks for initiating the conversation. I would also like to be part of this discussion, can you please also send me an invite too.

chinmayms commented:
@chrevanthreddy  Sure. 1:30 works. Thanks

chrevanthreddy commented:
@shkhrgpt Could you please where I can send you the invite? @chinmayms I have sent the invite to your Gmail Id.

shkhrgpt commented:
@chrevanthreddy Please send it to sgupta@pepperdata.com

akshayrai commented:
+1 LGTM

chinmayms commented:
@akshayrai  @shankar37 
As discussed, could we merge this with linkedin:master?

shkhrgpt commented:
+1 for merging.
